lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp

```diff
@@ -402,6 +420,36 @@ bool DebugNamesDWARFIndex::SameParentChain(
   return true;
 }
 
+bool DebugNamesDWARFIndex::WithinParentChain(
+    llvm::ArrayRef<llvm::StringRef> query_parent_names,
+    llvm::ArrayRef<DebugNames::Entry> parent_chain) const {
+  if (parent_chain.size() < query_parent_names.size())
+    return false;
+
+  size_t query_idx = 0, chain_idx = 0;
+  while (query_idx < query_parent_names.size() &&
+         chain_idx < parent_chain.size()) {
+    if (SameAsEntryATName(query_parent_names[query_idx],
+                          parent_chain[chain_idx])) {
+      ++query_idx;
+      ++chain_idx;
+    } else {
+      // Name does not match, try next parent_chain entry if the current entry
+      // is namespace because the current one can be an inline namespace.
```

Comment:
do we have a way of guaranteeing it *is*?
If the comparison that failed to match is with a real (non inline) namespace, we would return a false result here.

Reply:
Would that help? I don't see anything in the DIE that let's us identify inline namespaces...

https://godbolt.org/z/GT98Yfbqv

Reply:
Cool, did not know godbolt can show dwarfdump output.

`DWARFASTParserClang` identifies namespace DIE with `DW_AT_export_symbols` as inline namespace:
https://github.com/llvm/llvm-project/blob/81c3499531c3fe03827bd8bc890e3a16db9e4c3c/lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp#L3472-L3473

Reply:
Ok, good to know we have a mechanism to detect inline namespaces.

That still leaves the question of how to avoid false results though. I understand peeking at the DIE would be expensive for `dwo`s, but I don't think we can sacrifice the correctness of the answer for these queries. One possibility to explore is to add another IDX entry for namespaces, which would include whether they export symbols or not.

Reply:
The design is made to perform as much filtering as possible in `DebugNamesDWARFIndex::GetTypesWithQuery()`, it does not have to be accurate, because it will eventually delegate to base class `DWARFIndex::ProcessTypeDieMatchQuery()` which performs accurate semantics as the original callsite.